x/emissions/keeper/keeper.go

```diff
@@ -1139,44 +1161,205 @@ func (k *Keeper) SetDelegateStakeUponReputer(ctx context.Context, topicId TopicI
 	return k.stakeUponReputer.Set(ctx, key, stake)
 }
 
-// For a given topic id and reputer address, get their stake removal information
-func (k *Keeper) GetStakeRemovalByTopicAndAddress(ctx context.Context, topicId TopicId, address ActorId) (types.StakePlacement, error) {
-	key := collections.Join(topicId, address)
-	return k.stakeRemoval.Get(ctx, key)
-}
-
 // For a given address, adds their stake removal information to the removal queue for delay waiting
 // The topic used will be the topic set in the `removalInfo`
 // This completely overrides the existing stake removal
-func (k *Keeper) SetStakeRemoval(ctx context.Context, address ActorId, removalInfo types.StakePlacement) error {
-	key := collections.Join(removalInfo.TopicId, address)
-	return k.stakeRemoval.Set(ctx, key, removalInfo)
+func (k *Keeper) SetStakeRemoval(ctx context.Context, removalInfo types.StakeRemovalInfo) error {
+	byBlockKey := collections.Join3(removalInfo.BlockRemovalCompleted, removalInfo.TopicId, removalInfo.Reputer)
+	err := k.stakeRemovalsByBlock.Set(ctx, byBlockKey, removalInfo)
+	if err != nil {
+		return err
+	}
+	byActorKey := collections.Join3(removalInfo.Reputer, removalInfo.TopicId, removalInfo.BlockRemovalCompleted)
```

Comment:
This key is exactly the same as the above, if that was intentional, can you create one key and use it twice?

Or can you just set this data in one place instead of two?

Reply:
its also a little spooky to do it in a different order each time, what if there is a mistake?

Reply:
The order is what's important here, they're two different types of keys in the [IAVL store](https://github.com/cosmos/iavl/blob/master/docs/overview.md). In the first case you're looking up ("Alice", "Topic 1", "Block 356") and in the second case you're looking up ("Block 356", "Topic 1", "Alice").

My understanding of the system is basically that keys are concatenated together and then put into a tree data structure. So changing the order of the components that make up the key completely changes the resulting location in the tree - ("Alice", "Topic 1", "Block 356") may end up in the left most position in one tree and ("Block 356", "Topic 1", "Alice") will end up in the middle or something like that

We care about the ability to iterate over members of the data store so we need to be able to traverse the tree both by block and also by delegator/reputer. Hence the "Actor" and "Block" suffixes. I'd like to have just one data structure but I know of no efficient way to traverse the tree to iterate over the values in the data store without having to either go through every available block which has a removal in the case of looking up an actor, or iterate over every actor looking for a specific block.

This is akin to solidity having two mappings forwards and backwards to store the same data in an O(1) lookup indexed by different keys